src/monoclr/clrmod.c

```diff
@@ -1,70 +1,223 @@
+    MonoDomain *domain;
+    MonoAssembly *pr_assm;
+    MonoMethod *shutdown;
+    const char *pr_file;
+    char *error;
+    char *init_name;
+    char *shutdown_name;
+    PyObject *module;
+} PyNet_Args;
 
-    /* Create the module and add the functions */
-#if PY_MAJOR_VERSION >= 3
-    m = PyModule_Create(&clrdef);
-#else
-    m = Py_InitModule3("clr", clr_methods, clr_module_doc);
-#endif
-    if (m == NULL)
-        return NULL;
-    PyModule_AddObject(m, "facade", Py_True);
-    Py_INCREF(Py_True);
+PyNet_Args *PyNet_Init(void);
+static PyNet_Args *pn_args;
 
-    pn_args = PyNet_Init(1);
+PyMODINIT_FUNC
+PyInit_clr(void)
+{
+    pn_args = PyNet_Init();
     if (pn_args->error)
     {
         return NULL;
     }
 
-    if (NULL != pn_args->module)
-        return pn_args->module;
+    return pn_args->module;
+}
+
+void PyNet_Finalize(PyNet_Args *);
+void main_thread_handler(PyNet_Args *user_data);
+
+// initialize Mono and PythonNet
+PyNet_Args *PyNet_Init()
+{
+    PyObject *pn_module;
+    PyObject *pn_path;
+    PyNet_Args *pn_args;
+    pn_args = (PyNet_Args *)malloc(sizeof(PyNet_Args));
+
+    pn_module = PyImport_ImportModule("pythonnet");
+    if (pn_module == NULL)
+    {
+        pn_args->error = "Failed to import pythonnet";
+        return pn_args;
+    }
+
+    pn_path = PyObject_CallMethod(pn_module, "get_assembly_path", NULL);
+    if (pn_path == NULL)
+    {
+        Py_DecRef(pn_module);
+        pn_args->error = "Failed to get assembly path";
+        return pn_args;
+    }
 
-    return m;
+    pn_args->pr_file = PyUnicode_AsUTF8(pn_path);
+    pn_args->error = NULL;
+    pn_args->shutdown = NULL;
+    pn_args->module = NULL;
+
+#ifdef __linux__
+    // Force preload libmono-2.0 as global. Without this, on some systems
+    // symbols from libmono are not found by libmononative (which implements
+    // some of the System.* namespaces). Since the only happened on Linux so
+    // far, we hardcode the library name, load the symbols into the global
+    // namespace and leak the handle.
+    dlopen("libmono-2.0.so", RTLD_LAZY | RTLD_GLOBAL);
+#endif
+
+    pn_args->init_name = "Python.Runtime:InitExt()";
+    pn_args->shutdown_name = "Python.Runtime:Shutdown()";
+
+    pn_args->domain = mono_jit_init_version(MONO_DOMAIN, MONO_VERSION);
+
+    // XXX: Skip setting config for now, should be derived from pr_file
+    // mono_domain_set_config(pn_args->domain, ".", "Python.Runtime.dll.config");
+
+    /*
+     * Load the default Mono configuration file, this is needed
+     * if you are planning on using the dllmaps defined on the
+     * system configuration
+     */
+    mono_config_parse(NULL);
+
+    /* I can't use this call to run the main_thread_handler. The function
+     * runs it in another thread but *this* thread holds the Python
+     * import lock -> DEAD LOCK.
+     *
+     * mono_runtime_exec_managed_code(pn_args->domain, main_thread_handler,
+     *                                pn_args);
+     */
```

Comment:
I think you can safely remove this comment. The official Mono doc says directly executing code is the way forward: http://docs.go-mono.com/index.aspx?link=xhtml%3adeploy%2fmono-api-embedding.html (search for `mono_runtime_exec_managed_code` on the page)

Reply:
Yeah, I was planning on switching to clr-loader as the next step, so this one will be obsolete. I'll remove it anyhow.